### ISSUE TITLE
Flaky kernel compile failure on WinUWP bot

### ISSUE THREAD

**Issue by cbracken** [flutter/flutter#94492]
I'm seeing CI failures in the WinUWP presubmit shard that appear to be related to building kernel_blob for test fixtures. 

Example failures include:
* Dark SDK roll https://github.com/flutter/engine/pull/30033: [failure log](https://ci.chromium.org/ui/p/flutter/builders/try/Windows%20UWP%20Engine/8617/overview)
* Skia roll https://github.com/flutter/engine/pull/30034: [failure log](https://ci.chromium.org/ui/p/flutter/builders/try/Windows%20UWP%20Engine/8618/overview)
* Third-party patch Dark SDK roll https://github.com/flutter/engine/pull/30013: [failure log](https://ci.chromium.org/ui/p/flutter/builders/try/Windows%20UWP%20Engine/8638/overview)
* Fuchsia patch  https://github.com/flutter/engine/pull/30041: [failure log](https://ci.chromium.org/ui/p/flutter/builders/try/Windows%20UWP%20Engine/8645/overview)

The error seems to be related to kernel builds, specifically during kernel compiles during the step where GN invokes dart running frontend_server.dart.snapshot. Example:
```
FAILED: gen/flutter/runtime/assets/no_plugin_registrant_kernel_blob.bin 
C:/b/s/w/ir/cipd_bin_packages/cpython3/bin/python3.exe ../../build/gn_run_binary.py ../../flutter/prebuilts/windows-x64/dart-sdk/bin/dart.exe C:/b/s/w/ir/cache/builder/src/flutter/prebuilts/windows-x64/dart-sdk/bin/snapshots/frontend_server.dart.snapshot --sdk-root C:/b/s/w/ir/cache/builder/src/out/winuwp_debug/flutter_patched_sdk --target flutter --sound-null-safety --output-dill gen/flutter/runtime/assets/no_plugin_registrant_kernel_blob.bin --depfile C:/b/s/w/ir/cache/builder/src/out/winuwp_debug/gen/flutter/runtime/snapshot__dsk__ds_no_plugin_registrant.depfile.d C:/b/s/w/ir/cache/builder/src/flutter/runtime/fixtures/no_dart_plugin_registrant_test.dart
Traceback (most recent call last):
  File "../../build/gn_run_binary.py", line 23, in <module>
    subprocess.check_output(args, stderr=subprocess.STDOUT, universal_newlines=True)
  File "C:\b\s\w\ir\cipd_bin_packages\cpython3\bin\lib\subprocess.py", line 415, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "C:\b\s\w\ir\cipd_bin_packages\cpython3\bin\lib\subprocess.py", line 493, in run
    with Popen(*popenargs, **kwargs) as process:
  File "C:\b\s\w\ir\cipd_bin_packages\cpython3\bin\lib\subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\b\s\w\ir\cipd_bin_packages\cpython3\bin\lib\subprocess.py", line 1311, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] The system cannot find the file specified
```

**Comment by cbracken** [comment]
The engine tree is now closed due to the same flake happening in post-submit.

**Comment by CaseyHillers** [comment]
I suspect this may be flaking as it's missing an explicit dependency. The [flaky bot](https://chromium-swarm.appspot.com/bot?id=flutter-prod-windows-us-central1-c-0-7ngq) had not run a devicelab test, but the [successful bot](https://chromium-swarm.appspot.com/bot?id=flutter-prod-windows-us-central1-b-10-4gdd) had.

I created a task with `clobber: true` to wipe the deps and do a clean run of the task to see if it passed. Otherwise the failure is elsewhere - https://chromium-swarm.appspot.com/task?id=5790ff45f85d2b10

**Comment by CaseyHillers** [comment]
I found https://chromium-swarm.appspot.com/bot?id=flutter-try-windows-us-central1-b-18-2l93 which is a clean bot that ran the task on first try. I suspect there may be another task that impacted the environment

**Comment by CaseyHillers** [comment]
Lowering priority as this isn't directly blocking the tree.

**Comment by jmagman** [comment]
Tree closed again: https://luci-milo.appspot.com/p/flutter/builders/prod/Windows%20Host%20Engine/12909

**Comment by CaseyHillers** [comment]
Escalating to P0 as the try pool is indicating this build issue is happening very frequently - https://ci.chromium.org/p/flutter/builders/try/Windows%20UWP%20Engine?limit=100

**Comment by CaseyHillers** [comment]
@cbracken is able to reproduce locally. After manually extracting the dart sdk, this is fixed. I'm working on getting a windows cloudtop, but it'll take a day to get setup

A list of current thoughts:
1. The file has a write lock placed on it due to a running process, but Windows doesn't fail the extraction as it'll finish the write on reboot
2. https://github.com/flutter/engine/blob/main/tools/download_dart_sdk.py hiding failed to write on windows

**Comment by CaseyHillers** [comment]
@cbracken I'm going to hand this off to you as it sounds like you're making the most progress. I don't think my windows env will be ready for testing this until EOD.

**Comment by cbracken** [comment]
Unfortunately I'm already working on a very high priority item and won't be able to pick this up.

**Comment by christopherfujino** [comment]
Could this be related to https://github.com/flutter/engine/commit/e022c9283c4053a533599ab9b1e3559931248670 cc @zanderso

**Comment by CaseyHillers** [comment]
I'm blocked on admin privileges being granted to my windows cloud top. For now, I can't even clone a copy of the engine or run python. Worst case is approval will take 24 hours.

**Comment by CaseyHillers** [comment]
My current understanding is this can be reproduced by rebuilding on windows (or forcing the dart sdk to be updated).

To work on this, we'll need to add checks in https://github.com/flutter/engine/blob/main/tools/download_dart_sdk.py to ensure there's no write locks on `bin/` and if so forcefully remove those files.

I'm unassigning myself as this is marked as P0, and i'm unable to work on this until 12/6. However, if it's still open and unassigned i'll assign it back.

**Comment by CaseyHillers** [comment]
\cc @godofredoc

**Comment by godofredoc** [comment]
Checking out dart sdk was failing silently when downloading canvas kit and failing later during the process when trying to use the dart sdk:

https://ci.chromium.org/p/flutter/builders/try/Windows%20Host%20Engine/30055

**Comment by godofredoc** [comment]
This issue was mitigated setting this bug as P1

**Comment by godofredoc** [comment]
Actually getting back to P0 as it keeps failing very frequently on staging.

**Comment by chinmaygarde** [comment]
Is any progress being made on this? I haven't seen chatter about this issue in the sheriff chat. Does this need to be at a P0 still?

**Comment by godofredoc** [comment]
The remaining part is to patch https://github.com/flutter/engine/blob/main/tools/download_dart_sdk.py to handle readonly files properly.

https://stackoverflow.com/questions/2656322/shutil-rmtree-fails-on-windows-with-access-is-denied describes the fix. Anyone from engine that can help us land and test the change?

**Comment by godofredoc** [comment]
By the way the failures on presubmit were related to all PRs that needed to get rebased to pick up the change to update the flags of download_dart_sdk.py

**Comment by godofredoc** [comment]
This is fixed. Remaining failures in presubmit are because the PRs need to get rebased.
